crates/circuits/mod-builder/src/core_chip.rs

```diff
@@ -194,145 +191,178 @@ impl FieldExpressionCoreChip {
+            .adapter
+            .read(state.memory, instruction, adapter_row)
+            .into();
 
-        let outputs: Vec<BigUint> = self
-            .air
-            .output_indices()
-            .iter()
-            .map(|&i| vars[i].clone())
-            .collect();
-        let writes: Vec<F> = outputs
-            .iter()
-            .map(|x| biguint_to_limbs_vec(x.clone(), limb_bits, field_element_limbs))
-            .concat()
-            .into_iter()
-            .map(|x| F::from_canonical_u32(x))
-            .collect();
+        let (writes, inputs, flags) = run_field_expression(self, &data, instruction);
+
+        // TODO(arayi): Should move this to fill_trace_row
+        self.expr
+            .generate_subrow((self.range_checker.as_ref(), inputs, flags), core_row);
+
+        self.adapter
+            .write(state.memory, instruction, adapter_row, &writes.into());
 
-        let ctx = AdapterRuntimeContext::<_, DynAdapterInterface<_>>::without_pc(writes);
-        Ok((ctx.into(), FieldExpressionRecord { inputs, flags }))
+        *state.pc = state.pc.wrapping_add(DEFAULT_PC_STEP);
+        *trace_offset += width;
+        Ok(())
     }
 
-    fn get_opcode_name(&self, _opcode: usize) -> String {
-        self.name.clone()
+    fn fill_trace_row(&self, mem_helper: &MemoryAuxColsFactory<F>, row: &mut [F]) {
+        let (adapter_row, _) = row.split_at_mut(A::WIDTH);
+        self.adapter.fill_trace_row(mem_helper, (), adapter_row);
     }
+}
 
-    fn generate_trace_row(&self, row_slice: &mut [F], record: Self::Record) {
-        self.air.expr.generate_subrow(
-            (self.range_checker.as_ref(), record.inputs, record.flags),
-            row_slice,
-        );
+impl<F, A> StepExecutorE1<F> for FieldExpressionStep<A>
+where
+    F: PrimeField32,
+    A: 'static
+        + for<'a> AdapterExecutorE1<F, ReadData: Into<DynArray<u8>>, WriteData: From<DynArray<u8>>>,
+{
+    fn execute_e1<Mem, Ctx>(
+        &mut self,
+        state: VmStateMut<Mem, Ctx>,
+        instruction: &Instruction<F>,
+    ) -> Result<()>
+    where
+        Mem: GuestMemory,
+    {
+        let data: DynArray<_> = self.adapter.read(state.memory, instruction).into();
+
+        let writes = run_field_expression(self, &data, instruction).0;
+        self.adapter
+            .write(state.memory, instruction, &writes.into());
+        *state.pc = state.pc.wrapping_add(DEFAULT_PC_STEP);
+        Ok(())
     }
+}
 
-    fn air(&self) -> &Self::Air {
-        &self.air
+fn run_field_expression<F: PrimeField32, A>(
+    step: &FieldExpressionStep<A>,
+    data: &DynArray<u8>,
+    instruction: &Instruction<F>,
+) -> (DynArray<u8>, Vec<BigUint>, Vec<bool>) {
+    let field_element_limbs = step.expr.canonical_num_limbs();
+    let limb_bits = step.expr.canonical_limb_bits();
+
+    let data = data.0.iter().map(|&x| x as u32).collect_vec();
+
+    assert_eq!(data.len(), step.num_inputs() * field_element_limbs);
+
+    let mut inputs = vec![];
```

Comment:
`Vec::with_capacity`

also this entire function for sure needs to be optimized later